### PR TITLE
Restore default mode

### PR DESCRIPTION
**Pull request by comp-phys-marc**
**Context:** Restores default mode to `standard`. Clarifies docs regarding this. Also raises if probs are provided when a method is chosen that is incompatible with probs.

**Description of the Change:** Restores default. Adds checks and raises to `_quantum_automatic_differentiation`

**Benefits:** Cleaner UI/UX. Proper error messages will now be raised.

**Possible Drawbacks:** N/A

### CODE REVIEW

File: pennylane/gradients/hadamard_gradient.py [hunk 1]
```diff
@@ -608,6 +608,11 @@ def _quantum_automatic_differentiation(tape, trainable_param_idx, aux_wire) -> t
         trainable_op, _, _ = tape.get_operation(trainable_param_idx)
         _, generators = _get_pauli_generators(trainable_op)
 
+        if tape.measurements[0].obs is None:
+            raise ValueError(
+                "The circuit must have observables in order to use Quantum Automatic Differentiation."
```
Comment (mudit2812) [inline pennylane/gradients/hadamard_gradient.py]: Do we need to capitalize? Asking because I'm not sure if this is the standard we've been using to refer to the feature everywhere else as well.

Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: We have been capitalizing it everywhere :)

Comment (mudit2812) [reply pennylane/gradients/hadamard_gradient.py]: Thanks for confirming. In that case, I'm happy to approve. I see the PR is still in draft mode, so I'll wait for it to be ready for review and pass CI.

**Comment by mudit2812** [comment]
Is this PR still useful after the discussion in today's stand up?

**Comment by comp-phys-marc** [comment]
> Is this PR still useful after the discussion in today's stand up?

I am going to totally repurpose it.

**Comment by comp-phys-marc** [comment]
Ok I think this is ready for review again. Sorry for the big change in approach.

**Comment by comp-phys-marc** [comment]
Never mind, I need to take a different approach once more. XD

**Comment by comp-phys-marc** [comment]
Alright, I have removed the deprecation.

File: pennylane/gradients/hadamard_gradient.py [hunk 2]
```diff
@@ -115,14 +115,14 @@ def hadamard_grad(
...
         aux_wire (pennylane.wires.Wires): Auxiliary wire to be used for the Hadamard tests.
-            If ``None`` (the default), a suitable wire is inferred from the wires used in
-            the original circuit and ``device_wires``.
+            If ``None`` (the default) and ``mode`` is "standard" or "reversed", a suitable wire
+            is inferred from the wires used in the original circuit and ``device_wires``.
         device_wires (pennylane.wires.Wires): Wires of the device that are going to be used for the
             gradient. Facilitates finding a default for ``aux_wire`` if ``aux_wire`` is ``None``.
         mode (str): Specifies the gradient computation mode. Accepted values are
             ``"standard"``, ``"reversed"``, ``"direct"``, ``"reversed-direct"``, or ``"auto"``.
             The default ``"auto"`` chooses the method that leads to the
```
Comment (lillian542) [inline pennylane/gradients/hadamard_gradient.py]:
```suggestion
            The ``"auto"`` mode chooses the method that leads to the
```

File: pennylane/gradients/hadamard_gradient.py [hunk 3]
```diff
@@ -115,14 +115,14 @@ def hadamard_grad(
...
-            fewest total executions.
+            fewest total executions, given the wires available.
```
Comment (lillian542) [inline pennylane/gradients/hadamard_gradient.py]: Maybe something like this? "Available" is a bit ambiguous, could also mean "available on the device" or "passed to `device_wires`"

```suggestion
            fewest total executions, given the wire provided to ``aux_wires``.
```

Comment (lillian542) [reply pennylane/gradients/hadamard_gradient.py]: Or "based on the circuit observable and whether or not an auxiliary wire has been provided"? Maybe that's too verbose.

File: pennylane/gradients/hadamard_gradient.py [hunk 4]
```diff
@@ -158,18 +158,18 @@ def hadamard_grad(
...
     ...     qml.RY(params[1], wires=0)
     ...     qml.RX(params[2], wires=0)
     ...     return qml.expval(qml.Z(0)), qml.probs(wires=0)
     >>> params = jax.numpy.array([0.1, 0.2, 0.3])
     >>> jax.jacobian(circuit)(params)
-    (Array([-0.3875172 , -0.18884787, -0.38355704], dtype=float64),
-     Array([[-0.1937586 , -0.09442394, -0.19177852],
-            [ 0.1937586 ,  0.09442394,  0.19177852]], dtype=float64))
+    (Array([-0.3875172 , -0.18884787, -0.38355705], dtype=float32), Array([[-0.1937586 , -0.09442393, -0.19177853],
+           [ 0.1937586 ,  0.09442393,  0.19177853]], dtype=float32))
```
Comment (lillian542) [inline pennylane/gradients/hadamard_gradient.py]: I still get float64 - it's just a local config thing. I think we usually assume people are using float64.

File: pennylane/gradients/hadamard_gradient.py [hunk 5]
```diff
@@ -242,46 +242,29 @@ def hadamard_grad(
...
-        >>> gradient_kwargs = {"aux_wire": "c"}
-        >>> @qml.qnode(dev, interface="jax", diff_method="hadamard", gradient_kwargs=gradient_kwargs)
-        >>> def circuit(params):
-        ...    qml.RX(params[0], wires="a")
-        ...    qml.RY(params[1], wires="a")
-        ...    qml.RX(params[2], wires="a")
-        ...    return qml.expval(qml.Z("a"))
-        >>> params = jax.numpy.array([0.1, 0.2, 0.3])
-        >>> jax.jacobian(circuit)(params)
-        Array([-0.3875172 , -0.18884787, -0.38355704], dtype=float64)
```
Comment (lillian542) [inline pennylane/gradients/hadamard_gradient.py:284]: Should we put this back? It's technically still applicable until we do the deprecation/removal of using `None` to mean "infer wires for me". But it does feel like a pretty fringe application.

We can clarify "If you use custom wires on your device, and you want to use the "standard" or "reversed" modes, you need to pass an auxiliary wire."

File: pennylane/gradients/hadamard_gradient.py [hunk 6]
```diff
@@ -115,14 +115,14 @@ def hadamard_grad(
...
             the list of trainable parameters.
         aux_wire (pennylane.wires.Wires): Auxiliary wire to be used for the Hadamard tests.
-            If ``None`` (the default), a suitable wire is inferred from the wires used in
-            the original circuit and ``device_wires``.
+            If ``None`` (the default) and ``mode`` is "standard" or "reversed", a suitable wire
+            is inferred from the wires used in the original circuit and ``device_wires``.
         device_wires (pennylane.wires.Wires): Wires of the device that are going to be used for the
             gradient. Facilitates finding a default for ``aux_wire`` if ``aux_wire`` is ``None``.
         mode (str): Specifies the gradient computation mode. Accepted values are
             ``"standard"``, ``"reversed"``, ``"direct"``, ``"reversed-direct"``, or ``"auto"``.
```
Comment (lillian542) [inline pennylane/gradients/hadamard_gradient.py]:
```suggestion
        mode (str): Specifies the gradient computation mode. Accepted values are
            ``"standard"``, ``"reversed"``, ``"direct"``, ``"reversed-direct"``, or ``"auto"``. Defaults to ``"standard"``.
```

File: pennylane/gradients/hadamard_gradient.py [hunk 7]
```diff
@@ -632,11 +620,16 @@ def _count_groupings(paulis):
 
     if direct:
         if standard:
+            assert_no_probability(tape.measurements, "direct")
```
Comment (lillian542) [inline pennylane/gradients/hadamard_gradient.py]: I'm wondering if we could make the error message we raise here more actionable? Especially if the goal is to have "auto" become the default so that people don't have to think about it, I'm not sure if its immediately clear that the solution here is to provide an `aux_wire`. I'm wondering if we can catch this error and then raise something more informative from it, like "Automatically selected direct mode for differentiation because no auxiliary wire was passed. Computing the gradient of probabilities with the direct gradient transform is not supported. Consider explicitly supplying an aux_wire to the gradient method."

File: pennylane/gradients/hadamard_gradient.py [hunk 8]
```diff
@@ -632,11 +620,16 @@ def _count_groupings(paulis):
...
             return _direct_hadamard_test(tape, trainable_param_idx, aux_wire)
+
+        # The case where probs would appear here is covered on line 594 since probs are measurements but not obs
```
Comment (lillian542) [inline pennylane/gradients/hadamard_gradient.py]: Some probs measurements have an observable. You can ask for `qml.probs(op=qml.X(0)`, and that will not trigger the error in line 594.

File: pennylane/gradients/hadamard_gradient.py (hunk 3 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: I think the verbose version makes more sense

File: pennylane/gradients/hadamard_gradient.py (hunk 8 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: Oh shoot

File: pennylane/gradients/hadamard_gradient.py (hunk 7 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: Sure

File: pennylane/gradients/hadamard_gradient.py (hunk 5 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py:284]: I can put it back.

File: pennylane/gradients/hadamard_gradient.py (hunk 3 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: Added in https://github.com/PennyLaneAI/pennylane/pull/8875/commits/fef957e16ec117d474bc09f8ce9996edda35c365

File: pennylane/gradients/hadamard_gradient.py (hunk 5 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py:284]: Added back in https://github.com/PennyLaneAI/pennylane/pull/8875/commits/23fc28c421283e87df6c786d7cf85f250e00fea6

File: pennylane/gradients/hadamard_gradient.py (hunk 4 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: Ok... I will revert to float64

File: pennylane/gradients/hadamard_gradient.py (hunk 7 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: Only, it is a bit more complicated as the pressence of an `aux_wire` is not the only factor deciding whether we will use one method or the other, it also depends on the number of groupings of observables and generators.

Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: I can include the longer error message in the specific case where we are doing direct and not reversed-direct or reversed. But we will need a more generic error in the case of the other two.

File: pennylane/gradients/hadamard_gradient.py (hunk 8 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: Ok addressed in https://github.com/PennyLaneAI/pennylane/pull/8875/commits/54379056cc091ee80fee1bfbaa7d01a910164f1c

File: pennylane/gradients/hadamard_gradient.py (hunk 7 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: Let me know what you think https://github.com/PennyLaneAI/pennylane/commit/54379056cc091ee80fee1bfbaa7d01a910164f1c

File: pennylane/gradients/hadamard_gradient.py (hunk 8 above)
Comment (lillian542) [reply pennylane/gradients/hadamard_gradient.py]: Discussed offline and updated to default to `standard` if there are probs and an `aux_wire` is specified, and error out otherwise

File: pennylane/gradients/hadamard_gradient.py (hunk 7 above)
Comment (lillian542) [reply pennylane/gradients/hadamard_gradient.py]: Discussed offline and updated to default to standard if there are probs and an aux_wire is specified, and error out otherwise

File: tests/gradients/core/test_hadamard_gradient.py [hunk 9]
```diff
@@ -499,11 +499,26 @@ def test_automatic_mode_multiple_observables(self, mocker):
...
+        op = qml.evolve(qml.X(0) @ qml.X(1) + qml.Y(2) + qml.Z(0) @ qml.Z(1), t)
+        mp = qml.expval(qml.Z(0) @ qml.X(1) + qml.Y(0) + qml.X(0) @ qml.Z(1))
+        tape = qml.tape.QuantumScript([op], [mp, qml.probs((0, 1))])
+
+        with pytest.raises(
+            ValueError,
+            match="Computing the gradient of probabilities is only possible with the standard",
+        ):
+            _, _ = qml.gradients.hadamard_grad(tape, mode="auto")
+
```
Comment (lillian542) [inline tests/gradients/core/test_hadamard_gradient.py:517]: Can we throw a check in the same test to confirm that if we call it again, but with `aux_wire=2`, the error isn't raised, and a `mocker.spy` shows the standard Hadamard test being called?

Comment (comp-phys-marc) [reply tests/gradients/core/test_hadamard_gradient.py:517]: yep

Comment (comp-phys-marc) [reply tests/gradients/core/test_hadamard_gradient.py:517]: will have to be `aux_wire=3` though :)

File: pennylane/gradients/hadamard_gradient.py [hunk 10]
```diff
@@ -354,17 +352,17 @@ def hadamard_grad(
...
-        invovled in each problem to choose whether the standard or reversed order is preferred.
-        It also takes into account whether we have one or multiple measurements, and whether we have an
-        auxilliary wire.
+        total executions, given the wires available. Any auxilliary wires must be provided explicitly.
+        This method takes into account the number of observables and the number of generators invovled
+        in each problem to choose whether the standard or reversed order is preferred. It also takes
+        into account whether we have one or multiple measurements, and whether we have an auxilliary wire.
 
         ===============  ===============  ==============================
         Auxilliary Wire  Standard Order   Method
```
Comment (mudit2812) [inline pennylane/gradients/hadamard_gradient.py]:
```suggestion
        total executions, given the wires available. Any auxiliary wires must be provided explicitly.
        This method takes into account the number of observables and the number of generators involved
        in each problem to choose whether the standard or reversed order is preferred. It also takes
        into account whether we have one or multiple measurements, and whether we have an auxiliary wire.

        ===============  ===============  ==============================
        Auxiliary Wire  Standard Order   Method
```

File: pennylane/gradients/hadamard_gradient.py [hunk 11]
```diff
@@ -158,18 +159,18 @@ def hadamard_grad(
     to use during autodifferentiation:
 
     >>> import jax
+    >>> from pennylane import numpy as np
```
Comment (mudit2812) [inline pennylane/gradients/hadamard_gradient.py]: The example uses jax, so no need to import pennylane.numpy
```suggestion
```

File: pennylane/gradients/hadamard_gradient.py [hunk 12]
```diff
@@ -214,7 +215,7 @@ def hadamard_grad(
         ...     qml.RX(params[2], wires=0)
         ...     return qml.expval(qml.Z(0))
         >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
```
Comment (mudit2812) [inline pennylane/gradients/hadamard_gradient.py]:
```suggestion
        >>> params = qml.numpy.array([0.1, 0.2, 0.3], requires_grad=True)
```

File: pennylane/gradients/hadamard_gradient.py [hunk 13]
```diff
@@ -242,34 +243,31 @@ def hadamard_grad(
...
         ...     return qml.expval(qml.Z(0))
         >>> params = jax.numpy.array([0.1, 0.2, 0.3])
         >>> jax.jacobian(circuit)(params)
-        Array([-0.3875172 , -0.18884787, -0.38355704], dtype=float64)
-
-        If you use custom wires on your device, you need to pass an auxiliary wire.
+        Array([-0.3875172 , -0.18884787, -0.38355705], dtype=float64)
 
+        If you use custom wires on your device, and you want to use the "standard" or "reversed" modes, you need to pass an auxiliary wire.
         >>> dev_wires = ("a", "c")
```
Comment (mudit2812) [inline pennylane/gradients/hadamard_gradient.py:271]: This is rendering incorrectly because there is no new line
image

```suggestion
        If you use custom wires on your device, and you want to use the "standard" or "reversed" modes, you need to pass an auxiliary wire.

        >>> dev_wires = ("a", "c")
```

**Review by mudit2812 — COMMENTED** [review]
Looks good. I'll approve once the suggested changes are made and the rendered docs look good 😄

File: pennylane/gradients/hadamard_gradient.py (hunk 11 above)
Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: Actually without the import I was not able to run the exaple since it provides a flag to `np.array` saying that it is trainable. This param doesn't exist on usual `np`

Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py]: oh I see you changed it to refer to qml.numpy inline

Comment (mudit2812) [reply pennylane/gradients/hadamard_gradient.py]: Yeah, since the use of `pennylane.numpy` was limited to a later code block, it made sense to me to isolate it there, especially because the above code block is in the main docstring while the other code block is hidden in the "Usage details" section.

File: pennylane/gradients/hadamard_gradient.py [hunk 14]
```diff
@@ -213,8 +213,8 @@ def hadamard_grad(
         ...     qml.RY(params[1], wires=0)
         ...     qml.RX(params[2], wires=0)
         ...     return qml.expval(qml.Z(0))
-        >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
-        >>> qml.gradients.hadamard_grad(circuit)(params)
+        >>> params = qml.numpy.array([0.1, 0.2, 0.3], requires_grad=True)
```
Comment (gabrielasd) [inline pennylane/gradients/hadamard_gradient.py:216]: why use `qml.numpy` here?

Comment (comp-phys-marc) [reply pennylane/gradients/hadamard_gradient.py:216]: Because we need to have a trainable parameter to do machine learning. This is a gradient method. Normal numpy doesn't have a way for us to say this is a parameter we want to optimize over.

File: pennylane/gradients/hadamard_gradient.py [hunk 15]
```diff
@@ -602,6 +600,14 @@ def _quantum_automatic_differentiation(tape, trainable_param_idx, aux_wire) -> t
 
     direct = not aux_wire
 
+    if any(isinstance(m, ProbabilityMP) for m in tape.measurements):
+        if aux_wire:
+            return _hadamard_test(tape, trainable_param_idx, aux_wire)
+        raise ValueError(
+            "Computing the gradient of probabilities is only possible with the standard "
+            "Hadamard gradient, which requires an auxiliary wire. Please provide an aux_wire."
+        )
```
Comment (gabrielasd) [inline pennylane/gradients/hadamard_gradient.py]:
```suggestion
            "Hadamard gradient, which requires an auxiliary wire. Please provide a value for aux_wire."
        )
```

File: pennylane/gradients/hadamard_gradient.py (hunk 14 above)
Comment (gabrielasd) [reply pennylane/gradients/hadamard_gradient.py:216]: Thank you!
